File: packages/provide-uterm/src/provide/uterm/colors/mode.py
#
# SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
# SPDX-License-Identifier: AGPL-3.0-or-later
#
"""Unified color-mode dispatcher — handles str or bytes transparently."""

from __future__ import annotations

from typing import Literal, overload

from provide.uterm.colors.downgrade import downgrade_to_16, downgrade_to_256

ColorMode = Literal["passthrough", "256", "16"]


@overload
def apply_color_mode(data: bytes, mode: ColorMode) -> bytes: ...
@overload
def apply_color_mode(data: str, mode: ColorMode) -> str: ...


def apply_color_mode(data: bytes | str, mode: ColorMode) -> bytes | str:
    """Apply a color-mode filter to data, preserving the input type.

    Accepts either :class:`str` or :class:`bytes`. Bytes are decoded as
    latin-1 for the regex pass and re-encoded on return so stream filters
    (e.g. gateway writers) and text renderers share the same
    implementation.

    Args:
        data: ANSI text or raw bytes containing SGR sequences.
        mode:
            ``"passthrough"``: return data unchanged (zero-cost hot path).
            ``"256"``: downgrade truecolor SGR to xterm-256 cube.
            ``"16"``: downgrade truecolor SGR to the base 16-color palette.

    Returns:
        Filtered data, same type as the input.
    """
    if mode == "passthrough":
        return data
    if isinstance(data, bytes):
        text = data.decode("latin-1", errors="replace")
        text = downgrade_to_256(text) if mode == "256" else downgrade_to_16(text)
        return text.encode("latin-1", errors="replace")
    return downgrade_to_256(data) if mode == "256" else downgrade_to_16(data)


__all__ = ["ColorMode", "apply_color_mode"]

File: packages/provide-uterm-csharp/src/Provide.Uterm/Colors/Mode.cs
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

using System.Text;

namespace Provide.Uterm.Colors;

public enum ColorMode
{
    Passthrough,
    Mode256,
    Mode16,
}

public static class ColorModes
{
    public const string Passthrough = "passthrough";
    public const string Mode256 = "256";
    public const string Mode16 = "16";

    public static ColorMode Parse(string mode) => mode switch
    {
        Passthrough => ColorMode.Passthrough,
        Mode256 => ColorMode.Mode256,
        _ => ColorMode.Mode16,
    };

    public static string ApplyColorMode(string data, ColorMode mode) => mode switch
    {
        ColorMode.Passthrough => data,
        ColorMode.Mode256 => Downgrade.DowngradeTo256(data),
        _ => Downgrade.DowngradeTo16(data),
    };

    public static string ApplyColorMode(string data, string mode) => ApplyColorMode(data, Parse(mode));

    public static byte[] ApplyColorModeBytes(byte[] data, ColorMode mode)
    {
        if (mode == ColorMode.Passthrough)
        {
            return data;
        }

        var target = mode == ColorMode.Mode256 ? ColorMode.Mode256 : ColorMode.Mode16;
        // Latin-1 one-to-one: run SGR regex over string form of bytes.
        var latin1 = Encoding.GetEncoding("ISO-8859-1");
        var text = latin1.GetString(data);
        var rewritten = Sgr.SgrRegexp.Replace(text, m => Sgr.RewriteParams(m.Groups[1].Value, target));
        return latin1.GetBytes(rewritten);
    }

    public static byte[] ApplyColorModeBytes(byte[] data, string mode) =>
        ApplyColorModeBytes(data, Parse(mode));
}
